### PR TITLE
feat(app): Expose Labware Calibration Status on the FileInfo Page

### LINKED ISSUE #5555: Epic: Display(Make skipping) labware calibration easier
***Update based on Design Review*** 

We do not need tp create "Skip Labware Cal" as we believe it will cause confusion and open up potential issues down the road

**The new path forward is to display the labware calibration data for the labware that has a saved calibration already**

Users frequently are unsure when or if they can skip labware calibration, even though it's completely reasonable to skip labware calibration if everything is already calibrated.

This epic adds functionality to the protocol file info page to

1. check if all labware is already calibrated
1. display that calibration
1. add a new "proceed to run" CTA next to the "proceed to calibration" button

https://app.zeplin.io/project/5aa97729db58a2192f10d0d6/screen/5e458c93024bdcb402d46ec9

┆Issue is synchronized with this [Wrike Task](https://www.wrike.com/open.htm?id=535447675) by [Unito](https://www.unito.io/learn-more)

### PR DESCRIPTION
**Pull request by Laura-Danielle**
# Overview
Closes #5555 . This is the front-end side of skipping labware calibration. 

# Changelog

* Add epic/selector/action etc in order to access the new `labware/calibrations` endpoint
* Add tests associated with the above
* Modify the FileInfo component to match with the new design
* Add component tests

# Review requests
There are more endpoints under `labware/calibrations` to access a single labware calibration -- I wasn't sure if this was useful to implement at this time. Happy to add it to the pull request as needed (wouldn't be used in any of the components here).

I am also still working on some more component tests (having issues with mocks), but I wanted to get this in front of people. Let me know if there are other tests I should add.

## Test Plan

* [ ] Check that uploading a file uses the new design in #5555
* [ ] With no labware previously calibrated it should:
* [ ] Say `not yet calibrated`
* [ ] The button should still say `Proceed to Calibrate` and bring you to the calibration page
* [ ] With labware previously calibrated it should:
* [ ] Give the offset values for the labware
* [ ] The button should still say `Proceed to Run` and bring you to the run page
* [ ] Check that while calibrating, the calibration is updated in real time when you navigate back to the file info page

# Risk assessment
High, this is a new user facing feature that will be used very often. We should thoroughly test and go through user testing rigors. Should this potentially be behind a feature flag?

┆Issue is synchronized with this [Wrike Task](https://www.wrike.com/open.htm?id=535469622) by [Unito](https://www.unito.io/learn-more)

### CODE REVIEW

File: app/src/components/FileInfo/index.js [hunk 1]
```diff
@@ -10,6 +16,16 @@ import { Continue } from './Continue'
 import { UploadError } from '../UploadError'
 import styles from './styles.css'
 
+import { selectors as robotSelectors } from '../../robot'
+import {
+  selectors as labwareCalibrationSelectors,
+  actions as labwareActions,
+} from '../../calibration/labware'
```
Comment (mcous) [inline app/src/components/FileInfo/index.js:23]: `calibration` is a top-level module and components and other clients of that module should not (except in weird, circular dependency cases which aren't happening here) be reaching into the internal structure of that module.

`calibration/index.js` should reexport whatever is needed from labware internally so that client files can simply import from `calibration`

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+export function ProtocolLabwareList({
+  labware,
+  quantity,
+  calibration,
+}: ProtocolLabwareListProps): React.Node {
+  const [targetProps, tooltipProps] = useHoverTooltip({
+    placement: TOOLTIP_AUTO,
+  })
+  const iconComponent = (
+    <Icon name="information" className={styles.calibration_data_info_icon} />
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:37]: You can use style props here and avoid adding more CSS

```js
<Icon name="information" size={SIZE_1} color={C_MED_GRAY} />
```

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  )
+  const toolTipComponent = (
+    <Tooltip {...tooltipProps}>
+      {'calibrated offset from labware origin point'}
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:41]: Unnecessary braces and quotes

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    </Tooltip>
+  )
+
+  return (
+    <Flex
+      textAlign={DIRECTION_COLUMN}
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:47]: `textAlign` and `DIRECTION_COLUMN` don't go together. Did you mean `flexDirection={DIRECTION_COLUMN}` or `textAlign={TEXT_ALIGN_SOMETHING}` (where SOMETHING is LEFT, CENTER, RIGHT, or JUSTIFY)?

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      justifyContent={JUSTIFY_SPACE_BETWEEN}
+      fontSize={FONT_SIZE_BODY_1}
+      fontWeight={FONT_WEIGHT_REGULAR}
+      color={C_DARK_GRAY}
+    >
+      <SectionContentFlex title="Type" width={FLEX_AUTO}>
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:53]: `SPACING_AUTO` is a little more correct here than `FLEX_AUTO` here in terms of intent, but both constants are `auto` so it doesn't really matter. That being said, I think `width: auto` is just the default, so the `width` prop here might not be necessary

Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:53]: Can "Type" be in a constant instead of a magic string

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        {labware.map(name => (
+          <Text key={name}>{name}</Text>
+        ))}
+      </SectionContentFlex>
+      <SectionContentFlex title="Quantity" width={FLEX_AUTO}>
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:58]: Const instead of magic string and same comment about `width`

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        {quantity.map((item, index) => (
+          <Text key={`${index}`}>{item}</Text>
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:60]: I think the template string here might be unnecessary? Apologies if I'm wrong here; I forget what Flow likes as `key`s and what it complains about

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        ))}
+      </SectionContentFlex>
+      <SectionContentFlex
+        title="Calibration Data"
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:64]: Magic string

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        width={FLEX_AUTO}
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:65]: SPACING_AUTO but also maybe unnecessary?

File: app/src/calibration/api-types.js [hunk 10]
```diff
@@ -34,3 +34,38 @@ export type CalibrationStatus = {|
...
+|}
+
+export type TipLengthData = {|
+  value: number,
+  lastModified: string,
+|}
+
+export type CalibrationData = {|
+  offset: OffsetData,
+  tipLength: ?TipLengthData,
```
Comment (mcous) [inline app/src/calibration/api-types.js:50]: If possible, I've been trying to prefer explicit `TipLengthData | null` or `TipLengthData | void` to the more ambiguous `?TipLengthData`, which is equivalent to `TipLengthData | null | void`

File: app/src/calibration/api-types.js [hunk 11]
```diff
@@ -34,3 +34,38 @@ export type CalibrationStatus = {|
...
+|}
+
+export type LabwareCalibrationObjects = {|
+  attributes: SingleLabwareCalibration,
+  type: string,
+  id: string,
+|}
+
+export type AllLabwareCalibrations = {|
+  data: Array<?LabwareCalibrationObjects>,
```
Comment (mcous) [inline app/src/calibration/api-types.js:68]: Same comment: avoid `?Type` if possible

File: app/src/calibration/labware/__fixtures__/labware-calibration.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  LabwareCalibrationObjects,
+  AllLabwareCalibrations,
+} from '../../api-types'
+
+export const mockLabwareCalibration: LabwareCalibrationObjects = {
+  attributes: {
+    calibrationData: {
+      offset: {
+        value: [0.0, 0.0, 0.0],
+        lastModified: '2020-03-27 19:43:24.642318',
```
Comment (mcous) [inline app/src/calibration/labware/__fixtures__/labware-calibration.js:20]: Are these strings examples of what the API returns? If so, this is a problem for the API. All timestamps should be one of:

- MS since epoch
- ISO 8601 UTC strings

Or we're going to have a bad time. In my head I had thought we'd settled on ms since epoch a few years back.

File: app/src/calibration/labware/__tests__/selectors.test.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
+// @flow
+
+import * as Fixtures from '../__fixtures__'
+import * as StatusFixtures from '../../__fixtures__'
+import * as Selectors from '../selectors'
+
+import type { State } from '../../../types'
+
+describe('calibration selectors', () => {
+  describe('getCalibrationStatus', () => {
```
Comment (mcous) [inline app/src/calibration/labware/__tests__/selectors.test.js:10]: Copy paste typo

File: app/src/calibration/labware/actions.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+export const fetchAllLabwareCalibrations = (
+  robotName: string
+): Types.FetchLabwareCalibrationAction => ({
+  type: Constants.FETCH_ALL_LABWARE_CALIBRATIONS,
+  payload: { robotName },
+  meta: {},
+})
+
+export const fetchLabwareCalibrationSuccess = (
```
Comment (mcous) [inline app/src/calibration/labware/actions.js:20]: Nitpick, but should this (and params below) be calibration**s** rather than calibration to match the trigger action?

File: app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+import type {
+  ActionToRequestMapper,
+  ResponseToActionMapper,
+} from '../../../robot-api/operators'
+import type { Epic } from '../../../types'
+import type { FetchLabwareCalibrationAction } from '../types'
+
+const mapActionToRequest: ActionToRequestMapper<FetchLabwareCalibrationAction> = action => {
+  // $FlowFixMe(lc, 2020-07-08): lodash flow types do not support keys with values of different types
+  const payloadWithoutRobot = omitBy(action.payload, function(value, key) {
```
Comment (mcous) [inline app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js:20]: You can replace this with an object spread, which flow will understand:

```js
const { robotName, ...payloadWithoutRobot } = action.payload
```

File: app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    return key === 'robotName'
+  })
+  // $FlowFixMe(lc, 2020-07-08): lodash flow types do not support keys with values of different types
+  const queryDict = omitBy(payloadWithoutRobot, isEmpty)
```
Comment (mcous) [inline app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js:24]: This seems unnecessary; what's it supposed to be doing?

File: app/src/calibration/labware/epic/__tests__/fetchLabwareCalibrationsEpic.test.js [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+  runEpicTest,
+} from '../../../../robot-api/__utils__'
+import * as Fixtures from '../../__fixtures__'
+import * as Actions from '../../actions'
+import { labwareCalibrationEpic } from '..'
+
+const makeTriggerActionAllCalibrations = robotName =>
+  Actions.fetchAllLabwareCalibrations(robotName)
+
+describe('fetch labware calibration epics', () => {
```
Comment (mcous) [inline app/src/calibration/labware/epic/__tests__/fetchLabwareCalibrationsEpic.test.js:13]: We're missing test cases for calls with query parameters

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        icon={iconComponent}
+        toolTipComponent={toolTipComponent}
+        toolTipProps={targetProps}
+      >
+        <table style={{ borderSpacing: { SPACING_1 } }}>
+          <tbody>{calibration}</tbody>
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:71]: This is a little interesting because you have no real guarantee that whatever is passed in for the `calibration` prop is appropriate as a child of a `<table>`. It creates a little implicit dependency.

If possible, I would move the `<table>` to whatever is creating `calibration`

File: app/src/components/FileInfo/ProtocolLabwareList.js [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+        ))}
+      </SectionContentFlex>
+      <SectionContentFlex
+        title="Calibration Data"
+        width={FLEX_AUTO}
+        icon={iconComponent}
+        toolTipComponent={toolTipComponent}
+        toolTipProps={targetProps}
+      >
+        <table style={{ borderSpacing: { SPACING_1 } }}>
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareList.js:70]: Prefer `css` prop to inline styles:

```js
import { css } from 'styled-components'

<table css={
  css`
    border-spacing: ${SPACING_1}
  `
}>
```

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 20]
```diff
@@ -1,52 +1,61 @@
...
+  const labwareQuantity = Object.keys(labware).map(type => `x${labware[type]}`)
 
   return (
     <InfoSection title={TITLE}>
-      <LabwareTable>{labwareList}</LabwareTable>
+      <ProtocolLabwareList
+        labware={Object.keys(labware)}
+        quantity={labwareQuantity}
+        calibration={labwareCalibration}
+      ></ProtocolLabwareList>
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:58]: Tag can be self-closing here

```suggestion
      />
```

**Review by mcous — COMMENTED** [review]
This is looking like it's on the right track code-wise. Added some early feedback as you continue to fix up tests and stuff

File: app/src/calibration/labware/__fixtures__/labware-calibration.js (hunk 12 above)
Comment (Laura-Danielle) [reply app/src/calibration/labware/__fixtures__/labware-calibration.js:20]: It should be ISO 8601, I got the wrong format for this. I will change it.

File: app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js (hunk 16 above)
Comment (Laura-Danielle) [reply app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js:24]: If we do not take out the empty keys for the query, it will try to hit the endpoint with something to the effect of labware/calibrations?loadName=version=..etc which will cause an error

File: app/src/components/FileInfo/ProtocolLabwareList.js (hunk 2 above)
Comment (Laura-Danielle) [reply app/src/components/FileInfo/ProtocolLabwareList.js:37]: The color parameter is missing from Icon. Should I add it to the IconProps, or continue using the CSS?

File: app/src/components/FileInfo/ProtocolLabwareList.js (hunk 4 above)
Comment (Laura-Danielle) [reply app/src/components/FileInfo/ProtocolLabwareList.js:47]: I think I meant flex direction

File: app/src/components/FileInfo/ProtocolLabwareList.js (hunk 2 above)
Comment (mcous) [reply app/src/components/FileInfo/ProtocolLabwareList.js:37]: It was added in https://github.com/Opentrons/opentrons/pull/5999. It looks like you started this branch before that PR was merged, so you should be able to use it if you sync up with `edge`

File: app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js (hunk 16 above)
Comment (mcous) [reply app/src/calibration/labware/epic/fetchAllLabwareCalibrationEpic.js:24]: - This functionality isn't used in this PR, so there either needs to be tests for it and comments about how it will be used in the future or it should be removed
- If the API cannot handle an empty string for a given query param then that is a bug in the API
- If we want to say `''`, `null`, and `undefined` should not be serialized to query parameters (which is reasonable), then that logic belongs in `app/src/robot-api/http.js` and `app/src/robot-api/__tests__/http.test.js`

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 21]
```diff
@@ -1,52 +1,78 @@
...
+export function ProtocolLabwareCard({
+  labware,
+  labwareCalibrations,
+}: ProtocolLabwareProps): React.Node {
+  if (Object.keys(labware).length === 0) return null
 
-function ProtocolLabwareCardComponent(props: Props) {
-  const { labware } = props
+  const labwareToParentMap = {}
+  Object.keys(labware).forEach((type, index) => {
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:34]: If an object were appropriate (it's not for the reason above), this should be a `reduce`, not a `forEach`. But, since it should be an array, you should be using `Array.map`

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 22]
```diff
@@ -1,52 +1,78 @@
...
+    const currentLabware = labwareCalibrations[type]?.attributes
+    let calibrationData
+    if (currentLabware) {
+      const offset = currentLabware?.calibrationData?.offset.value
+      const X = parseFloat(round(offset[0], 1)).toFixed(1)
+      const Y = parseFloat(round(offset[1], 1)).toFixed(1)
+      const Z = parseFloat(round(offset[2], 1)).toFixed(1)
+      calibrationData = (
+        <Flex
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:43]: I think you should move all these components / presentational concerns into `ProtocolLabwareList.js`, which is already dealing with setting up the table to look correct. `ProtocolLabwareCard.js` can deal with the data concerns and then pass an array of `{x, y, z}` objects down to `ProtocolLabwareList`

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 23]
```diff
@@ -1,52 +1,78 @@
...
-  const labwareCount = countBy(labware, 'type')
-  const labwareList = Object.keys(labwareCount).map(type => (
-    <tr key={type}>
-      <td>{type}</td>
-      <td>{`x${labwareCount[type]}`}</td>
-    </tr>
-  ))
+    return (labwareToParentMap[type] = {
+      parent: labware[type].parent,
+      quantity: `x${labware[type].count}`,
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:66]: This `x` is a presentational concern; I think this component should stick to simply passing down the quantity as a number

File: app/src/components/FileInfo/index.js [hunk 24]
```diff
@@ -22,19 +37,69 @@ export type FileInfoProps = {|
...
+  }, [dispatch, robotName])
   let uploadError = props.uploadError
 
+  const labware = useSelector((state: State) =>
+    robotSelectors.getLabware(state)
+  )
+  const modulesBySlot = useSelector((state: State) =>
+    robotSelectors.getModulesBySlot(state)
+  )
+  console.log(modulesBySlot)
```
Comment (mcous) [inline app/src/components/FileInfo/index.js:54]: Remove console.log

File: app/src/components/FileInfo/index.js [hunk 25]
```diff
@@ -22,19 +37,69 @@ export type FileInfoProps = {|
...
+  })
+
   return (
     <div className={styles.file_info_container}>
       <InformationCard />
       <ProtocolPipettesCard robotName={robot.name} />
       <ProtocolModulesCard robot={robot} />
-      <ProtocolLabwareCard />
+      <ProtocolLabwareCard
+        labware={labwareDisplayNames}
```
Comment (mcous) [inline app/src/components/FileInfo/index.js:100]: All the logic that builds up `labwareDisplayNames` and `calibrationLoadNamesMap` is too complicated to live in a component's render function. It should be moved to a selector to build up the single array of data that you need.

I also recommend that you move that selector down into `ProtocolLabwareCard`, so that `ProtocolLabwareCard` can simply take a `robotName` prop to feed the selector. Then you end up with a component that looks like:

- `FileInfo` - template component to arrange cards on the page
  - `ProtocolLabwareCard` - "smart" component to collect labware info from state and organize it
    - `ProtocolLabwareList` - "presentation" component to display data in a table

The tests then look like:

- Selector to get labware calibration data - standard selector tests that make sure the correct array(s) pop out given the state and robotName input
- `FileInfo` - test that it renders a `ProtocolLabwareCard` component and hands it the correct robotName
- `ProtocolLabwareCard` - test that it calls the selector properly and passes the selector output to `ProtocolLabwareList`
- `ProtocolLabwareList` - test that it renders a table and a table row per item in the labware array and smoke test that the data from each item is rendered to HTML in some way

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 26]
```diff
@@ -1,52 +1,78 @@
...
 
-type SP = {| labware: Array<Labware> |}
+import { InfoSection } from './InfoSection'
+import { ProtocolLabwareList } from './ProtocolLabwareList'
+import type { LabwareCalibrationObjects } from '../../calibration'
 
-type Props = {| ...SP, dispatch: Dispatch |}
+export type ProtocolLabwareProps = {|
+  labware: {
+    [key: string]: { count: number, display: string, parent: string },
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:20]: I don't think an object is the best data type to select for this prop, but if it was, `key` should be replaced with something more descriptive. Since `key` is generic, I don't know from reading this if `key` is the loadName, the displayName, something else, or if the `key` in the `labware` map is the same as `key` in the `labwareCalibrations` map below

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 27]
```diff
@@ -1,52 +1,78 @@
...
+  },
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:21]: If objects were the best move here (which they're probably not), with Flow we want all object type definitions to be either explicitly exact: `{| foo: string |}`, where no other properties are allowed; or explicitely inexact: `{ foo: string, ... }`, where other properties are allowed.

Because of quirks in Flow, for maps (`[keyName: string]: ValueType`), explicitely inexact with an ellipses `...` is usually easier to work with and what I'd usually recommend.

File: app/src/components/FileInfo/index.js [hunk 28]
```diff
@@ -22,19 +37,69 @@ export type FileInfoProps = {|
 |}
 
 export function FileInfo(props: FileInfoProps): React.Node {
+  const dispatch = useDispatch<Dispatch>()
   const { robot, sessionLoaded, sessionHasSteps } = props
+  const { name: robotName } = robot
+  React.useEffect(() => {
+    dispatch(labwareFunctions.fetchAllLabwareCalibrations(robotName))
```
Comment (mcous) [inline app/src/components/FileInfo/index.js:44]: This side effect might make more sense to run in an epic that is triggered by a successful protocol upload as opposed to hooking it into this component mounting to the UI

File: app/src/components/FileInfo/ProtocolLabwareCard.js [hunk 29]
```diff
@@ -1,52 +1,78 @@
...
-  if (labware.length === 0) return null
-
-  const labwareCount = countBy(labware, 'type')
-  const labwareList = Object.keys(labwareCount).map(type => (
-    <tr key={type}>
-      <td>{type}</td>
-      <td>{`x${labwareCount[type]}`}</td>
-    </tr>
-  ))
+    return (labwareToParentMap[type] = {
```
Comment (mcous) [inline app/src/components/FileInfo/ProtocolLabwareCard.js:64]: This looks like the crux of your whole data model, and I think what you should have a selector to return.

```js
type LabwareSummaryItem = {|
  parent: string,
  quantity: number,
  displayName: string,
  calibration: {| x: number, y: number, z: number |} | null,
|}

const getProtocolLabwareSummary: (state: State, robotName: string) => Array<LabwareSummaryItem> = createSelector(
  // ... compose other selectors here that are currently composed in UI by FileInfo
)
```

Then, `ProtocolLabwareCard` becomes more like:

```js
export type ProtocolLabwareProps = {|
  robotName: string
|}

export function ProtocolLabwareCard(props: ProtocolLabwareProps): React.Node {
  const { robotName } = props
  const labwareList = useSelector((state: State) => {
    return getProtocolLabwareSummary(state, robotName)
  })

  return (
    <InfoSection title={TITLE}>
      <ProtocolLabwareList labwareList={labwareList} />
    </InfoSection>
  )
}
```

**Review by mcous — DISMISSED** [review]
Let me know if you have any questions about any of this

File: app/src/components/FileInfo/index.js (hunk 28 above)
Comment (Laura-Danielle) [reply app/src/components/FileInfo/index.js:44]: The reason we're not doing that is to allow for updates as a user calibrated their labware in the calibration page.

They might navigate back to the file info component, and the information should be as up to date as possible.

File: app/src/calibration/labware/__fixtures__/labware-calibration.js [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+    parent: 'fake_id',
+  },
+  id: 'some id',
+  type: 'Labware Calibration',
+}
+
+export const mockAllLabwareCalibraton: AllLabwareCalibrations = {
+  data: [mockLabwareCalibration],
+  meta: {},
+  type: 'Labware Calibration',
```
[the remark this answers is not in the document]
Comment (Laura-Danielle) [reply app/src/calibration/labware/__fixtures__/labware-calibration.js:39]: You're right, it has no type field when it's in a collection format. I incorrectly typed this.

File: app/src/components/FileInfo/__tests__/index.test.js [hunk 31]
```diff
@@ -0,0 +1,51 @@
+// @flow
```
Comment (mcous) [inline app/src/components/FileInfo/__tests__/index.test.js:1]: Avoid `index.test.js`; prefer naming it after the component it's testing (in this case, `FileInfo.test.js`)

File: app/src/components/FileInfo/__tests__/continue.test.js [hunk 32]
```diff
@@ -0,0 +1,92 @@
+// @flow
```
Comment (mcous) [inline app/src/components/FileInfo/__tests__/continue.test.js:1]: Improper filename casing

File: app/src/components/FileInfo/__tests__/labwarelist.test.js [hunk 33]
```diff
@@ -0,0 +1,76 @@
+// @flow
```
Comment (mcous) [inline app/src/components/FileInfo/__tests__/labwarelist.test.js:1]: Improper filename casing

File: app/src/epic.js [hunk 34]
```diff
@@ -16,7 +16,7 @@ import { shellEpic } from './shell/epic'
 import { alertsEpic } from './alerts/epic'
 import { systemInfoEpic } from './system-info/epic'
 import { sessionsEpic } from './sessions/epic'
-import { calibrationEpic } from './calibration/epic'
+import { calibrationEpic } from './calibration/'
```
Comment (mcous) [inline app/src/epic.js:19]: This change should be reverted

File: app/src/protocol/selectors.js [hunk 35]
```diff
@@ -1,7 +1,8 @@
 // @flow
+import { createSelector } from 'reselect'
```
Comment (mcous) [inline app/src/protocol/selectors.js:2]: Why was this import moved? `path` is a Node.js built-in library, and by convention, we put those first in the import list

File: app/src/robot-api/__tests__/http.test.js [hunk 36]
```diff
@@ -80,6 +80,15 @@ describe('robot-api http client', () => {
     )
   })
 
+  it('removes any empty query params', () => {
+    const url = robotApiUrl(robot, {
+      method: GET,
+      path: '/health',
+      query: { fakeParam: '' },
```
Comment (mcous) [inline app/src/robot-api/__tests__/http.test.js:87]: Missing test conditions for `null` and `undefined`. We should also probably add a `false` to make sure `false` _isn't_ removed, because that's a common bug in these sorts of utilities

File: app/src/robot-api/types.js [hunk 37]
```diff
@@ -14,7 +14,7 @@ export type RobotApiRequestOptions = {|
   method: Method,
   path: string,
   body?: { ... },
-  query?: { [param: string]: string | boolean | number },
+  query?: { [param: string]: string | boolean | number | null },
```
Comment (mcous) [inline app/src/robot-api/types.js:17]: Our runtime logic permits `undefined` here, so it's probably good to be flexible with the types and tack on a `void`. Also needs a `...` to mark the object as inexact

File: robot-server/robot_server/service/labware/router.py [hunk 38]
```diff
@@ -36,10 +36,13 @@ def _format_calibrations(
...
             value=tip_cal.value,
             lastModified=tip_cal.last_modified)
+        # TODO (lc, 07-13-2020) Once we categorize calibrations
+        # per slot, we should ensure that the parent displayed
+        # is the slot number.
         if calInfo.parent.module:
             parent_info = calInfo.parent.module
         else:
-            parent_info = calInfo.parent.slot
+            parent_info = ''
```
Comment (mcous) [inline robot-server/robot_server/service/labware/router.py:45]: Why an empty string over `None`?

**Review by mcous — COMMENTED** [review]
Not sure this is ready yet

File: robot-server/robot_server/service/labware/router.py (hunk 38 above)
Comment (sfoster1) [reply robot-server/robot_server/service/labware/router.py:45]: It's more accurate, I think. It's not that there's no parent - that just how we characterize the parent of "any slot".

Comment (sfoster1) [reply robot-server/robot_server/service/labware/router.py:45]: I'm gonna get this out of this PR anyway - I didn't see this PR had this in it and actually made #6211 to address it

**Comment by mcous** [comment]
Summary of changes that I pushed:

- Removed query param behavior from GET /labware/calibrations actions
    - The reducer wasn't built to handle partial responses, so I removed the functionality rather than build out the reducer further
    - This functionality of the actions wasn't used, so it wasn't needed
- Fixed bugs in the calibration reducer + selectors + typedefs where they wouldn't be able to handle having `GET /labware/calibrations` response in state without also having `GET /calibration/status` in state already 
- Reworked the selector that feeds the labware list UI
    - Renamed it to `getProtocolLabwareList`
    - Fixed incorrect usage of `Array.map`
    - Added missing logic to use namespace and version in labware identity check 
    - Added missing test cases to ensure proper flattening into the `quantity` count and proper linkage to calibration data from the response
- Reworked the labware info components to lean more heavily on the selector and make better use of primitive components

**Review by nusrat813 — APPROVED** [review]
Tested - working as intended

File: app/src/calibration/labware/selectors.js [hunk 39]
```diff
@@ -0,0 +1,97 @@
...
+  loadName: string,
+  namespace: string | null,
+  version: number | null,
+  parent: ModuleModel | null,
+  legacy: boolean,
+|}
+
+// TODO(mc, 2020-07-27): this selector should move to a protocol-focused module
+// when we don't have to rely on RPC-state selectors for protocol equipment info
+// NOTE(mc, 2020-07-27): due to how these endpo
```
Comment (b-cooper) [inline app/src/calibration/labware/selectors.js:38]: This comment was truncated somehow? Should it be removed?

File: app/src/calibration/labware/selectors.js [hunk 40]
```diff
@@ -0,0 +1,97 @@
...
+
+    return uniqueLabware.map(lw => {
+      const { definition: def, loadName, namespace, version, parent } = lw
+      const displayName = def ? getLabwareDisplayName(def) : loadName
+      const parentDisplayName = parent ? getModuleDisplayName(parent) : null
+      const matchesLabwareIdentity = target =>
+        target.loadName === loadName &&
+        target.namespace === namespace &&
+        target.version === version &&
+        (!parent || target.parent === parent)
```
Comment (b-cooper) [inline app/src/calibration/labware/selectors.js:75]: If `!parent` is supposed to matching the case where `parent === ''`, it would be clearer to write that check explicitly

File: app/src/components/FileInfo/Continue.js [hunk 41]
```diff
@@ -1,33 +1,48 @@
...
+  TOOLTIP_LEFT,
+} from '@opentrons/components'
+
+// TODO(mc, 2020-07-27): i18n
+const PROCEED_TO_CALIBRATE = 'Proceed to Calibrate'
+const VERIFY_CALIBRATIONS = 'Verify pipette and labware calibrations'
 
 export function Continue(): React.Node {
   const { path, disabledReason } = useSelector(getCalibrateLocation)
+  const [targetProps, tooltipProps] = useHoverTooltip({
```
Comment (b-cooper) [inline app/src/components/FileInfo/Continue.js:28]: 🛸 🔨 💁‍♂️

**Review by b-cooper — APPROVED** [review]
Looks good 🖌️

File: app/src/calibration/labware/selectors.js (hunk 39 above)
Comment (mcous) [reply app/src/calibration/labware/selectors.js:38]: Oops, I should finish that comment. It was adding a note that v1 labware will always return "no calibration data" and that needs to be accounted for in UI

File: app/src/calibration/labware/selectors.js (hunk 40 above)
Comment (mcous) [reply app/src/calibration/labware/selectors.js:75]: It is checking the case where `parent === null` (`modulesBySlot[slot].model` can never be `''` to my knowledge), but good call on the explicit check, regardless
